futures-select-macro/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+                CaseKind::Normal(pat, expr)
+            };
+
+            // `=> <expr>`
+            input.parse::<Token![=>]>()?;
+            let expr = input.parse::<Expr>()?;
+
+            // Commas after the expression are only optional if it's a `Block`
+            // or it is the last branch in the `match`.
+            let is_block = match expr { Expr::Block(_) => true, _ => false };
+            if is_block || input.is_empty() {
+                input.parse::<Option<Token![,]>>()?;
+            } else {
+                input.parse::<Token![,]>()?;
+            }
+
+            match case_kind {
+                CaseKind::Complete => select.complete = Some(expr),
+                CaseKind::Default => select.default = Some(expr),
+                CaseKind::Normal(pat, fut_expr) => {
+                    select.normal_fut_exprs.push(fut_expr);
+                    select.normal_fut_handlers.push((pat, expr));
+                },
+            }
+        }
+
+        Ok(select)
+    }
+}
+
+// Enum over all the cases in which the `select!` waiting has completed and the result
+// can be processed.
+//
+// `enum __PrivResult<_1, _2, ...> { _1(_1), _2(_2), ..., Complete }`
+fn declare_result_enum(
+    result_ident: Ident,
+    variants: usize,
+    complete: bool,
+    span: Span
+) -> (Vec<Ident>, syn::ItemEnum) {
+    // "_0", "_1", "_2"
+    let variant_names: Vec<Ident> =
+        (0..variants)
+            .map(|num| Ident::new(&format!("_{}", num), span))
+            .collect();
+
+    let type_parameters = &variant_names;
+    let variants = &variant_names;
+
+    let complete_variant = if complete {
+        Some(quote!(Complete))
+    } else {
+        None
+    };
+
+    let enum_item = parse_quote! {
+        enum #result_ident<#(#type_parameters,)*> {
+            #(
+                #variants(#type_parameters),
+            )*
+            #complete_variant
+        }
+    };
+
+    (variant_names, enum_item)
+}
+
+/// The `select!` macro.
+#[proc_macro_hack]
+pub fn select(input: TokenStream) -> TokenStream {
+    let parsed = syn::parse_macro_input!(input as Select);
+
+    let futures_crate: syn::Path = parsed.futures_crate_path.unwrap_or_else(|| parse_quote!(::futures_util));
+    let rand_crate: syn::Path = parse_quote!(#futures_crate::rand_reexport);
+
+    // should be def_site, but that's unstable
+    let span = Span::call_site();
+
+    let enum_ident = Ident::new("__PrivResult", span);
+
+    let (variant_names, enum_item) = declare_result_enum(
+        enum_ident.clone(),
+        parsed.normal_fut_exprs.len(),
+        parsed.complete.is_some(),
+        span,
+    );
+
+    // bind non-`Ident` future exprs w/ `let`
+    let mut future_let_bindings = Vec::with_capacity(parsed.normal_fut_exprs.len());
+    let bound_future_names: Vec<_> = parsed.normal_fut_exprs.into_iter()
+        .zip(variant_names.iter())
+        .map(|(expr, variant_name)| {
+            match expr {
```

Comment:
Since working with a `!Unpin` futures is a bit painful at the moment (must be manually pinned before use in `select`): Can this happen automatically here?
Maybe in a restricted form only when the `Future` is an rvalue and guaranteed to not be used outside of `select`?
I have no idea whether this can be detected, but it sounds like it could simplify some things.

Reply:
Similar question for `Future`s which are not fused.
I guess if one wants to use the same future for each iteration in a loop then it still needs to be pinned and fused manually, but if it's a new `Future` for each iteration it might be possible to do this automatically.

Reply:
> For loops, I think select! could somehow consume the future and do that automatically as well
I think for the situations where that's ok, one would mostly be fine by creating the future as an rvalue in this expression.

However there are situations where one really needs to operate on exactly the same future instance between loop iterations: Some `Future`s, like for the mutex and manualresetevent one, carry internal state between poll() calls, and destroying and recreating them would reset this. For those types it would e.g. mean they might never be able to acquire the mutex when getting recreated instead of staying the same - or at least the fairness guarantee would be different.

Same with using a `async fn` which contains multiple `await!`s as a Future in of the expressions: If it's recreated in each iteration, it will would always start running from the beginning. If it's kept between iterations, it keeps also it's progress. Automatically hoisting this out of the loop would be interesting, but then the concern of this macro would go beyond the poor `select!` part.

Reply:
I want to leave any functional changes to `select!` out of this change, so that it's just refactoring. If there are other features you'd like to suggest, I'd be happy to review them as separate PRs.